src/Workspaces/Core/Portable/Utilities/IComplarableHelper.cs

```diff
@@ -0,0 +1,34 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Roslyn.Utilities
+{
+    internal static class IComparableHelper
+    {
+        public static int CompareTo<T>(T item1, T item2, Func<T, IEnumerable<IComparable>> comparingComponentsMethod)
```

Comment:
i don't think we can do the allocations here.  an i think we should have hte immutbale array of IComparers.  But there should be an overload that allows you to pass state in.